Create initial Haskell basics/basics page (#7)
haskellfoundation/HaskellSchool: en/lessons/basics/basics.md

## Getting Started

### Installing Haskell

The recommended installation instructions for each OS can be found at
[ghcup](https://www.haskell.org/ghcup/).

You can check the version of the Haskell compiler installed by running
`ghc --version`:

    % ghc --version

>     The Glorious Glasgow Haskell Compilation System, version {{ site.ghc.version }}

Author: Currently site.ghc.version is `8.10.4`, but the new `ghci>` prompt is used.

### Trying Interactive Mode

Your installation should come with `ghci`, an interactive shell that allows you
to try out code in real time.

Follow this tutorial by running `ghci`:

>     GHCi, version {{ site.ghc.version }}: https://www.haskell.org/ghc/  :? for help

Author: This exceeds the 80 col limit.

Author: Perhaps this is better left like this.

>     ghci>

Author: This prompt is used for consistency with `ghci` and other pages, but `>` was also considered in the documentation-task-force office hours.

> To use `Text` in Haskell, type in:

Reviewer: ```suggestion
To use the `Text` type as in examples below, type in:
```

Author: Would ```To use the `Text` type in this tutorial, type in:``` be fine?

Reviewer: Sure.

```haskell
ghci> :set -XOverloadedStrings
ghci> import qualified Data.Text as T

> ```

Reviewer: I don't buy the `Text` type for an introductory material on today's Haskell: it adds a heck lot of fuss like all those type annotations -- they will look very confusing to a novice. Someday, maybe when something like [this proposal](https://github.com/ghc-proposals/ghc-proposals/pull/409) happens, I could change my mind.

Author: I'm not sure I want to introduce a datatype to beginners that I wouldn't recommend using. However, I can remove a lot of the type annotations by using functions such as `T.length` and `T.putStrLn`. 

@Kleidukos: what should I do here?

Reviewer: I don't understand this anxiety about `String`: if you don't write a text processing application (text as in books, novels, etc.), and, say, simply need to print a greeting to a user, why wouldn't you use `String`? Everything should be judged by the cost/benefit ratio: if your program uses just several short strings, there's no need to pull in `Text`. Note that `Char` (and therefore `String`) in Haskell is [Unicode-aware](https://www.haskell.org/onlinereport/haskell2010/haskellch6.html#x13-1190006.1.2). So we're only really trading performance here. Is this tutorial the right place to optimize for performance, or maybe we'd want to prioritize usability?

There would be a place to discuss all intricacies of all string types, I hope: `text` is just a part of the story, and somewhere we'd need to introduce ByteStrings...

Reviewer 2: > if you don't write a text processing application (text as in books, novels, etc.), and, say, simply need to print a greeting to a user, why wouldn't you use String?

As a general-purpose textual data container, String is very bad. String has a very poor API, performances that go against what people know in other languages except C and the old Erlang strings, and I am afraid that if their utility disappears once we go beyond printing a greeting to a user, they shouldn't be mentioned at all.

I think linked lists should be taught as its own subject, and it should be mentioned that a common idiom is to have ad-hoc textual data represented as a list of `Char`, and outlining that it may be a nice data structure to implement templates that must be changed frequently[¹](https://samuelmullen.com/articles/phoenix-templates/).

> if your program uses just several short strings, there's no need to pull in Text.

I frequently uses several short strings and `Text` is an absolute must-have because I have _no wish_ to replicate its API on `String`. This is a matter of ergonomics and developer experience, not just data locality in contiguous slots of memory.

Reviewer 2: > However, I can remove a lot of the type annotations by using functions such as T.length and T.putStrLn.

@TrueBoxGuy Let us do so then

You can try out Haskell by typing in a few simple expressions:

```haskell
ghci> 2 + 3
5
ghci> 2 + 3 == 5
True
ghci> T.length "The quick brown fox jumps over the lazy dog"
43
```

> Perhaps you're able to figure out what some of these expressions mean.

Reviewer: ```suggestion
Perhaps you are able to figure out what some of these expressions mean.
```

## Basic Data Types

### Integers

Haskell has a bounded integer type called `Int`:

```haskell
ghci> 255 :: Int
255
```

> Note: specify the type to use with `::`.

Author: This reads as if the reader must commit some action when the note merely provides information about type annotations.

Haskell also supports arbitrary precision integers with `Integer`:

```haskell
ghci> 123456789101112131415
123456789101112131415
```
Note: `ghci` automatically defaults to `Integer` here.

### Floats

Haskell supports `Double`, a double-precision floating-point number.
At least one digit must come before the decimal point and Haskell supports
scientific `e` notation.

```haskell
ghci> 3.14
3.14
ghci> .14

> <interactive>:2:1: error: parse error on input ‘.’

Author: This doesn't highlight correctly on the generated page. Furthermore, the one line gap inserted (which `ghci` does because a parse error is generated) may be confusing for the reader to parse: line gaps are often used to separate sections of code.

Reviewer 2: Can you show me how it is rendered?

Author: This ![image](https://user-images.githubusercontent.com/17554188/120399138-531ba080-c333-11eb-8379-649b33d149ed.png)
rather than this
![image](https://user-images.githubusercontent.com/17554188/120399074-3bdcb300-c333-11eb-99db-6724c0383885.png)

ghci> 5e-10
5.0e-10
ghci> 2.7e10
2.7e10
```

### Booleans

Booleans in Haskell are represented with `True` and `False`:

```haskell
ghci> True
True
ghci> False
False
```

### Text

> Haskell supports unicode text. Strings are wrapped in double quotes:

Author: Mentioning both `Text` and overloaded strings without explanation may confuse the reader. Could a better word be found?

> Haskell supports unicode text. Strings are wrapped in double quotes:

Reviewer: ```suggestion
Haskell supports Unicode text. Strings are wrapped in double quotes:
```

```haskell
ghci> "Hello" :: T.Text
"Hello"
ghci> "jalapeño" :: T.Text
"jalape\241o"

> ```

Author: The constant type annotations may be confusing. A potential solution would be to use `Data.Text.IO`, but the extra import may be confusing; this solution would also enable for one to view unicode and escape codes working.

> "jalape\241o"

Reviewer: You may want to add that if the user wants the actual Unicode characters printed back (instead of the codepoints), they can use `putStrLn ("jalapeño" :: T.Text)`.

Author: I could go even further and print within the repl: 
>The constant type annotations may be confusing. A potential solution would be to use Data.Text.IO, but the extra import may be confusing; this solution would also enable for one to view unicode and escape codes working.

https://github.com/haskellfoundation/HaskellSchool/pull/7#discussion_r641925925

Do others agree with adding the cost of `Data.Text.IO`, and, furthermore, printing within the repl?

Special characters can be represented with numeric escapes or escape codes:

```haskell
ghci> "'hi\"" :: T.Text
"'hi\""
ghci> "\n" :: T.Text
"\n"
ghci> "\42" :: T.Text
"*"
```

You'll learn more about data types in [collections](../collections/) and
[functions](../functions/).

## Basic Operations

### Arithmetic

Haskell supports the basic operators `+`, `-`, `*`, and `/`.
Note that `/` never returns an integer.

```haskell
ghci> 2 + 2
4
ghci> 2 - 1
1
ghci> 2 * 5
10
ghci> 10 / 5
2.0

> ```

Reviewer: I'd add `^` and `**` which are surprisingly less known (in my experience) but are a must in today's data-centric world.

Reviewer 2: Agreed

For integer division and integer modulus, use `div` and `mod`:

```haskell
ghci> -8 `div` 3
-2
ghci> 243 `mod` 5
3
ghci> 2 * (5 `div` 2) + (5 `mod` 2)
5
```

### Boolean

Haskell provides the boolean operators `||`, `&&`, and `not` that operate on

> `Bool`:

Reviewer: Maybe worth adding a note on the usual short-circuiting behavior (Haskell is unusual in so many ways, so that this may come as a surprise).

```haskell
ghci> False || False
False
ghci> True || False
True

ghci> False && True
False
ghci> True && True
True

ghci> not True
False
ghci> not False
True
```

### Comparison

Haskell comes with comparison operators for ordering: `<=`, `>=`, `<`, and `>`.
The operator `==` compares equality whilst `/=` compares inequality.

> Values compared must be of the same type.

Author: These sentences could have a bit more cohesion. The introduction of equality and inequality seems abrupt.

Reviewer 3: > The operator `==` compares equality whilst `/=` compares inequality.

I'd consider rewording this. While I can catch the text's drift, a literal reading makes it sound equalities are being compared, rather than two values being compared to test for equality.

Maybe rewrite to:
> The operator `==` tests for equality whilst `/=` tests for inequality.
or
> The operator `==` checks for equality whilst `/=` checks for inequality.
---
> Values compared must be of the same type.
The phrasing seems a little bit awkward there. Consider:
> (The) compared values must have the same type.

dis
jointed
ghci> T.putStrLn "\42"
*
```

You'll learn more about data types in [collections](../collections/) and
[functions](../functions/).

## Basic Operations

### Arithmetic

Haskell supports the basic operators: `+`, `-`, `*`, `/`, `**` and `^`.
Note that `/` and `**` never operate on integers,
whereas `^` only accepts positive integer powers.

```console?lang=haskell&prompt=ghci>,ghci|
ghci> 2 + 2
4
ghci> 2 - 1
1
ghci> 2 * 5
10
ghci> 10 / 5
2.0
ghci> 2 ** (-0.5)
0.7071067811865476
ghci> 1.5 ^ 2
2.25
ghci> 5 ^ 2
25
```

For integer division and integer modulus, use `div` and `mod`:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> -8 `div` 3
-2
ghci> 243 `mod` 5
3
ghci> 2 * (5 `div` 2) + (5 `mod` 2)
5
```

### Boolean

Haskell provides the boolean operators `||`, `&&`, and `not` that operate on
`Bool`. These operators do not evaluate the second term if possible.

```console?lang=haskell&prompt=ghci>,ghci|
ghci> False || False
False
ghci> True || False
True

ghci> False && True
False
ghci> True && True
True

ghci> not True
False
ghci> not False
True
```

### Comparison

Haskell comes with comparison operators for ordering: `<=`, `>=`, `<`, and `>`.

> Conversely, the operator `==` compares equality whilst `/=` compares inequality.

Reviewer 3: ```suggestion
Naturally the operator `==` compares equality whilst `/=` compares inequality.
```
"Conversely" means "in the converse manner", and I don't think the statement that follows it is the converse of the statement that precedes it.

You could replace it with "As you might expect" however while most users of other languages will expect `==` to check for equality, most languages don't use `/=` for inequality and it is ambiguous whether the "As you might expect" applies to both clauses.

"Accordingly" seems _more_ fitting than "conversely", but not entirely fitting either.
"Naturally" works, though it does have an effect on the tone, so feel free to use your own judgement.

Author: I do want something which has a meaning close to conversely: I'm introducing equality operators now and want to highlight that contrast (that I'm covering the other side of the comparison operators). 

I feel like the word I want is "moreover": it highlights both that it's a new concept and that it adds on to what I've already said about comparison operators. What do you think?

> Conversely, the operator `==` compares equality whilst `/=` compares inequality.

Reviewer 3: > the operator `==` compares equality

Reading this literally makes it seem like the `==` operator compares equalities, rather than comparing values to see if they are equal.

You could replace "compares" with either "checks for" or "tests for".
```suggestion
Conversely, the operator `==`  checks for equality whilst `/=` checks for inequality.
```

Author: Yep. My changes didn't really solve https://github.com/haskellfoundation/HaskellSchool/pull/7#discussion_r641926100

> Values compared must be of the same type.

Reviewer 3: > Values compared must be of the same type.

This reads a little awkwardly in my opinion. Consider rewording:
```suggestion
Compared values must have the same type.
```
On the other hand, to avoid a misinterpretation this might be preferable:
```suggestion
Values compared using these operators must have the same type.
```

Author: I like the second one quite a bit.
